panel_client/keyboard_panel.py

```diff
@@ -30,50 +70,62 @@
                 'nipple': 'Octo bite raven girl nipple!',
                 'mouth': 'Octo kiss raven girl mouth!'
             },
-            'type': 'switch'
-        },
+            'type': 'button'
+        }
     ],
     [{
         'id': 'Froomulator',
         'state': '0',
+
+        # This is a shorthand form of `states`, where the first value (the array) contains the values for
+        # `state`, and the second value (the string) is a "template action": the actual actions for each
+        # state will be formed by replacing "%s" with the state.
         'actions': [['0', '1', '2'], 'Set Froomulator to %s!']
     }]
 ]
 
 
-def poll_keyboard(stop_queue, input_queue, stdin):
+def poll_keyboard(stop_event, input_queue, stdin):
   while True:
-
-    try:
-      _ = stop_queue.get(block=False)
+    if stop_event.is_set():
       return
-    except Queue.Empty:
-      pass
-    try:
-      user_input = stdin.readline()[:-1]
-      input_queue.put(user_input)
-    except EOFError:
-      pass
+    ready, _, __ = select.select([stdin], [], [], MIN_LATENCY_MS/1000)
+    if ready:
+      try:
+        user_input = stdin.readline()[:-1]
+        input_queue.put(user_input)
+      except EOFError:
+        pass
+
 
-    
 class KeyboardPanel(PanelStateBase):
   """Simulate a panel using the keyboard."""
 
   def __init__(self, stdin, player_number):
-    self.controls = {}
     self.input_queue = Queue.Queue()
-    self.stop_queue = Queue.Queue()
+    self.stop_event = threading.Event()
+    self.stop_event.clear()
+
     self.kbd_thread = threading.Thread(
-        target=poll_keyboard,
-        args=(self.stop_queue, self.input_queue, stdin))
+        target=poll_keyboard, args=(self.stop_event, self.input_queue, stdin))
     self.controls = CONTROL_SCHEMES[player_number - 1]
     self.display_message('Conrol scheme: {}'.format(self.controls))
     self.kbd_thread.start()
-    
+
+  def is_button(self, item):
+    matching_actions = [
+        action for action in self.controls
+        if action['id'] == item and action.get('type', '') == 'button'
+    ]
+    if matching_actions:
+      return True
+
   def get_state_updates(self):
     """Returns an iterable of control_id, state pairs for new user input."""
     try:
       item, action = self.input_queue.get(block=False).split(' ')
+      if self.is_button(item):
+        self.input_queue.put((item + ' 0'))
```

Comment:
Please replicate the HACK comment from `panel.py` here. The real panels won't need to do this, they'll emit a state change to `0` or `nothing` or whatever when the button is _actually_ depressed.

Reply:
Done